cgroups/fs/apply_raw.go

```diff
@@ -88,6 +91,93 @@ func Apply(c *cgroups.Cgroup, pid int) (map[string]string, error) {
 	return paths, nil
 }
 
+// Stop stops all processes in the current cgroup
+func Stop(c *cgroups.Cgroup) error {
+	d, err := getCgroupData(c, 0)
+	if err != nil {
+		return err
+	}
+
+	// first we send SIGTERM to all tasks inside the cgroup
+	if err := Kill(c, syscall.SIGTERM); err != nil {
+		return err
+	}
+
+	// if failed to stop with only SIGTERM, we send SIGKILL
+	if err := waitStop(d, 100*time.Millisecond, 5*time.Second); err != nil {
+		if err := Kill(c, syscall.SIGKILL); err != nil {
+			return err
+		}
+		return waitStop(d, 100*time.Millisecond, -1*time.Second)
```

Comment:
Hmm, right. if the process is stuck somehow, this wont return, and the cgroup will not be cleaned up. I think that the kernel will eventually try to clean everything when the container's main process exit and the PID namespace is destroyed, but i'm not too sure.

I think probably we should just return here after sending SIGKILL without waiting ? It will probably result in an immediate failure in cleaning up cgroups after ( but we retry the removal up to 5 times so hopefully it will be cleaned up eventually ). Worst case is that you have a dangling cgroup + an uninterrupted process until the PID namespace is removed.